### ISSUE TITLE
Customized optimizer in Keras

### ISSUE THREAD

**Issue by skchao74** [keras-team/keras#12987]
I am experimenting a new optimizer using Keras. I tried to imitate the code of Adagrad as shown in a previous question on [StackOverflow](https://stackoverflow.com/questions/41787873/how-adagrad-wroks-in-keras-what-does-self-weights-mean-in-keras-optimizer); in particular, besides the arguments are different, I have only slightly modified the update of "accumulators" and "parameters". The following is a part of my code:

```
def __init__(self, lr=0.01, c=0., mu=0.7, **kwargs):
        super(SKD, self).__init__(**kwargs)
        with K.name_scope(self.__class__.__name__):
            self.iterations = K.variable(0, dtype='int64', name='iterations')
            self.lr = K.variable(lr, name='lr') # lr
            self.mu = K.variable(mu, name='mu') # mu
            self.c = K.variable(c, name='c') # c
            
    @interfaces.legacy_get_updates_support
    def get_updates(self, loss, params):
        grads = self.get_gradients(loss, params)
        shapes = [K.int_shape(p) for p in params]   
        accumulators = [K.zeros(shape) for shape in shapes] 
        self.updates = [K.update_add(self.iterations, 1)]
        self.weights = accumulators 

        lr = self.lr
        mu = self.mu
        c = self.c
        l1 = c * K.pow(lr, 0.5 + mu) * K.pow(K.cast(self.iterations, K.floatx()), mu)
        for p, g, a in zip(params, grads, accumulators):
            new_a = a - lr * g 
            self.updates.append(K.update(a, new_a))
            new_p = np.sign(new_v) * (K.abs(new_v)-l1) 
            self.updates.append(K.update(p, new_p))
        return self.updates
````

The code does not do any update as the training loss hardly changes over iterations. To try to debug, I replace the update of new_p in the 3rd line from the bottom by 
`new_p = new_a`
This makes the algorithm updates equivalent to SGD, but the codes still do not decrease the training error. Since this code is very short and simple, after checking a million times I still do not know what I have done wrong. Any comments will be highly appreciated!
